src/charm.py

```diff
@@ -103,7 +125,123 @@ def _on_stop(self, _: ops.StopEvent) -> None:
     def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
         """Handle upgrade-charm."""
         self.unit.status = ops.MaintenanceStatus("Upgrading dependencies")
-        self.bind.prepare(self.unit.name)
+        self.bind.setup(self.unit.name)
+
+    def _on_leader_elected(self, _: ops.LeaderElectedEvent) -> None:
+        """Handle leader-elected event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _on_peer_relation_departed(self, _: ops.RelationDepartedEvent) -> None:
```

Comment:
This code made me think...From the doc, all units receive the relation departed event, including the departing one. But I don't know what would happen if the unit departing is also the leader.
I would suggest to, just in case, check if the unit is the departing one and return.

https://juju.is/docs/sdk/relation-name-relation-departed-event

Reply:
Good point !